src/lib.rs

```diff
@@ -1,70 +1,103 @@
+                rgba + (p_a[0] as i64 - p_b[0] as i64).abs()
         })
 }
 
+fn subimage_compare(image_a: &DynamicImage, image_b: &DynamicImage, x: u32, y: u32 ) -> i64 {
+    let sub_image_a = image_a.crop_imm(x,y,16,16).to_rgb();
+    let sub_image_b = image_b.crop_imm(x,y,16,16).to_rgb();
+    measure_dist_chunks(&sub_image_a, &sub_image_b)
+}
+
 pub fn generate_image(
-    image_buffer: &[u8],
+    image_buffer: &DynamicImage,
     iterations: u64,
     save_progress: bool,
     path: &Path,
 ) -> Result<Vec<u8>, ImageError> {
+    let now = Instant::now();
+
     let mut emoji_cache = emoji::EmojiCache::new();
-    let orig = load_from_memory(image_buffer)?.into_rgba();
-    let (width, height) = orig.dimensions();
-    let mut new_img = RgbaImage::new(width, height);
-    let mut dist = sums_chunked(&orig, &new_img);
 
-    for _ in 0..iterations {
+    let image_buffer_rgb = image_buffer.clone().to_rgb();
+    let (width, height) = image_buffer_rgb.dimensions();
+    let canvas_size = width * height;
+    let mut rng = thread_rng();
+    let mut new_img = DynamicImage::new_rgb16(width, height);
+    println!("image is {} by {} pixels", width, height);
+    let mut dist = measure_dist_chunks(&image_buffer_rgb, &new_img.to_rgb());
+    println!("dist is {}", dist);
+
+    let mut placed_count = 0;
+    for _  in 0..canvas_size/20 {
         let e = emoji_cache.get_emoji();
-        let x: u32 = random::<u32>() % width;
-        let y: u32 = random::<u32>() % height;
+        let x: u32 = (1..width).choose(&mut rng).unwrap();
+        let y: u32 = (1..height).choose(&mut rng).unwrap();
+        overlay(&mut new_img, e, x, y);
+        placed_count= placed_count+1;
+    }
+
+    dist = measure_dist_chunks(&image_buffer_rgb, &new_img.to_rgb());
+    println!("dist is {}", dist);
+    
+    for index in 0..iterations {
+        let e = emoji_cache.get_emoji();
+        let x: u32 = (1..width).choose(&mut rng).unwrap();
+        let y: u32 = (1..height).choose(&mut rng).unwrap();
         let mut temp_img = new_img.clone();
+        let temp_dist1 = subimage_compare(&image_buffer, &temp_img, x, y);
         overlay(&mut temp_img, e, x, y);
-        let temp_dist = sums_chunked(&orig, &temp_img);
-        if dist > temp_dist {
+        let temp_dist2 = subimage_compare(&image_buffer, &temp_img, x, y);
+        if temp_dist1 > temp_dist2 {
             new_img = temp_img;
-            dist = temp_dist;
-            if save_progress {
-                new_img.save(path)?;
-            }
+            placed_count= placed_count+1;
+        }
+        if index%1000==0 {
+            dist = measure_dist_chunks(&image_buffer_rgb, &new_img.to_rgb());
+            println!("iteration: {}, dist: {}, time: {}.{}, emoji: {}",
+                index,
+                dist,
+                now.elapsed().as_secs(),
+                now.elapsed().subsec_millis(),
+                placed_count
+                )
+        }
```

Comment:
i would delete this and also maybe just delete `dist` itself since we're not really using it anymore